Tessa/more flexible button v2

tesk9 opened the pull request:
Alternate version of https://github.com/NoRedInk/noredink-ui/pull/312.

This version of the Button module follows an attribute-setting pattern.

@avh4 @stoeffel @BrianHicks -- thoughts?

avh4 commented on src/Nri/Ui/Button/V9.elm:
```diff
@@ -0,0 +1,887 @@
+
+
+{-| -}
+type alias Attribute msg =
```
🎨 I think this should be a custom type instead of an alias, to keep this detail private.  Could be `type Attribute msg = Attribute (ButtonOrLink msg -> ButtonOrLink msg)` for now

stoeffel replied:
I agree with AV. I think this should be an opaque type.

avh4 commented on src/Nri/Ui/Button/V9.elm:
```diff
@@ -0,0 +1,887 @@
+    | Danger
+    | Premium
+      -- The following styles cannot be set directly.
+      -- Use a helper like `unfulfilled` to change the state instead.
```
🎨 I think it would be clearer to have two distinct type, one for the 4 "styles", and then a separate one with all 8 "color schemes" if a type for the scheme is really even necessary

avh4 replied:
(or maybe the view code could just destructure the Style and the State at the same time and then having the extra "schemes" as custom type variants wouldn't even be necessary?)

avh4 commented on src/Nri/Ui/Button/V9.elm:
```diff
@@ -0,0 +1,887 @@
+        , state = Enabled
+        , icon = Nothing
+        , customAttributes = []
+        }
```
💭 I think the documentation somewhere should list what the defaults are (medium, primary, widthUnbounded, enabled)

avh4 commented on src/Nri/Ui/Button/V9.elm:
```diff
@@ -0,0 +1,887 @@
+        , label = config.label
+        , state = config.state
+        , icon = config.icon
+        , customAttributes = config.customAttributes
```
❓ 🎨  was there a reason not to use record update syntax here and in `onClick`?

tesk9 replied:
uhhhhhhhhhhhhhhhh nope 😅

avh4 commented on src/Nri/Ui/Button/V9.elm:
```diff
@@ -0,0 +1,887 @@
+
+-}
+linkExternalWithTracking : Attribute msg
+linkExternalWithTracking (ButtonOrLink config) =
```
🎨 for some of these extra link attributes, I'm not sure I understand what all are used for, but I think some for instance will only be used when there's also an href and/or a msg -- in cases like those, I think it'd be nice to have the attribute function handle multiple private options at once, like `linkExternalWithTracking : Url -> msg -> Attribute msg` (which sets onClick, href, and linkType all at once)

tesk9 replied:
I thought this was a great suggestion and I started it, but I ran into this:

```
linkWithMethod : String -> String -> Attribute msg
linkWithMethod method url (ButtonOrLink config) =
    ButtonOrLink { config | linkType = WithMethod method, url = url }
```

We don't have the Url type defined in noredink-ui. I wouldn't want the Url data type to be versioned/seems like a pain to add it.

I guess I could make this a record? What do you think?

avh4 replied:
Hmm, yeah, I guess a record would be fine then.  Also I think just having two strings and documenting the order of them in the doc comment would also probably be fine 🤷‍♀

tesk9 replied:
All done!

avh4 commented on src/Nri/Ui/Button/V9.elm:
```diff
@@ -0,0 +1,887 @@
+
+
+{-| -}
+custom : Html.Attribute msg -> Attribute msg
```
🎨 I wonder if `custom` should just take a list of `Html.Attribute msg`, just to make the calling code a little less redundant in the case that there are multiple custom attributes?

tesk9 replied:
Yeah I thought about that and waffled! I can do multiple 👍

avh4 commented on src/Nri/Ui/Button/V9.elm:
```diff
@@ -0,0 +1,887 @@
+
+
+{-| -}
+icon : Svg -> Attribute msg
```
💭 I think we'd want to enforce that icon buttons always also have a label to use for alt text, so maybe icon should be `icon : String -> Svg -> Attribute msg` and set both the label and the icon at once?

tesk9 replied:
I think actually we don't want to label icon -- the icon's purpose is descriptive, and when that's the case the alt text should be explicitly set to `""`. 

That said, I don't have any idea about how to enforce that  here 🤔

avh4 replied:
Oh, right, there's still text when there's an icon.  That makes sense!

Hmm, maybe label actually shouldn't be an optional Attribute, then, and should just be a required string parameter to the button functions?

tesk9 replied:
An idea that occurred to me is to let the button take html children, like a normal button:

```
Button.button 
    [  Button.primary
    ,  Button.onClick DoAThing
    ]
    [  Html.text "My label"
    ]
```

Or maybe:

```
Button.button 
    [  Button.primary
    ,  Button.onClick DoAThing
    ]
    [  Button.icon myIcon
    ,  Button.label "HELLO!!!"
    ]
```

tesk9 replied:
Yeah I think required string sounds good..

avh4 replied:
I'd be hesitant about having it take a list of children because I suspect we didn't really think through making the CSS handling arbitrary children -- things like wrapping, vertical alignment, containing inline vs inline-block vs block elements, dealing with things of unexpected height, etc would all need to be handled if we allowed more flexibility in specifying the children.

tesk9 commented:
Alright! I think I've addressed all the feedback. I'll plan to merge this early next week, unless folks need or want more time to review. Thank you!!! 💚

stoeffel commented:
Looks great! Thanks for taking the time to make this so much nicer!

I agree with AV, that `Attribute` should be private.
